 # Amazon Reviews Sentiment Analysis

**Internship**  : CodeAlpha

**Name**   : Mehwish Iqbal

**Task** _03  :Sentiment Analysis



## **Project Overview**

This project focuses on **Sentiment Analysis and Emotion Detection** using Amazon customer reviews. The goal is to classify reviews as **Positive, Negative, or Neutral** and identify specific emotions expressed by customers.

The analysis also explores sentiment patterns across **customer ratings, countries, and review dates** to understand customer opinions and generate useful business insights.

## **Objectives**

* Classify customer reviews as **Positive, Negative, or Neutral**.
* Apply **NLP techniques and a keyword-based emotion lexicon** to detect emotions.
* Analyze sentiment patterns in **Amazon reviews**.
* Explore sentiment across **customer ratings and countries**.
* Identify sentiment trends over time.
* Generate insights for **marketing, product development, and customer experience**.


In [142]:
import pandas as pd
import numpy as np
import re
import nltk
import plotly.express as px
import seaborn as sns
import matplotlib.pyplot as plt

In [143]:
nltk.download('vader_lexicon')
nltk.download('punkt')

[nltk_data] Downloading package vader_lexicon to C:\Users\AL HAFIZ
[nltk_data]     COMPUTER\AppData\Roaming\nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!
[nltk_data] Downloading package punkt to C:\Users\AL HAFIZ
[nltk_data]     COMPUTER\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [144]:
df = pd.read_csv("Amazon_Reviews.csv", engine="python")

In [145]:
df.head()

,Reviewer Name,Profile Link,Country,Review Count,Review Date,Rating,Review Title,Review Text,Date of Experience
0,Eugene ath,/users/66e8185ff1598352d6b3701a,US,1 review,2024-09-16T13:44:26.000Z,Rated 1 out of 5 stars,A Store That Doesn't Want to Sell Anything,"I registered on the website, tried to order a ...","September 16, 2024"
1,Daniel ohalloran,/users/5d75e460200c1f6a6373648c,GB,9 reviews,2024-09-16T18:26:46.000Z,Rated 1 out of 5 stars,Had multiple orders one turned up and…,Had multiple orders one turned up and driver h...,"September 16, 2024"
2,p fisher,/users/546cfcf1000064000197b88f,GB,90 reviews,2024-09-16T21:47:39.000Z,Rated 1 out of 5 stars,I informed these reprobates,I informed these reprobates that I WOULD NOT B...,"September 16, 2024"
3,Greg Dunn,/users/62c35cdbacc0ea0012ccaffa,AU,5 reviews,2024-09-17T07:15:49.000Z,Rated 1 out of 5 stars,Advertise one price then increase it on website,I have bought from Amazon before and no proble...,"September 17, 2024"
4,Sheila Hannah,/users/5ddbe429478d88251550610e,GB,8 reviews,2024-09-16T18:37:17.000Z,Rated 1 out of 5 stars,If I could give a lower rate I would,If I could give a lower rate I would! I cancel...,"September 16, 2024"


In [146]:
df.shape

(21214, 9)

In [147]:
df.columns

Index(['Reviewer Name', 'Profile Link', 'Country', 'Review Count',
       'Review Date', 'Rating', 'Review Title', 'Review Text',
       'Date of Experience'],
      dtype='object')

## Step 3 — Dataset Overview

Before applying sentiment analysis, we inspect the dataset structure and check for missing and duplicate values.

### 3.1 Dataset Information

In [148]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21214 entries, 0 to 21213
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Reviewer Name       21214 non-null  object
 1   Profile Link        21163 non-null  object
 2   Country             21054 non-null  object
 3   Review Count        21055 non-null  object
 4   Review Date         21055 non-null  object
 5   Rating              21055 non-null  object
 6   Review Title        21055 non-null  object
 7   Review Text         21055 non-null  object
 8   Date of Experience  20947 non-null  object
dtypes: object(9)
memory usage: 1.5+ MB


### 3.2 Missing Values

In [149]:
df.isnull().sum()

Reviewer Name           0
Profile Link           51
Country               160
Review Count          159
Review Date           159
Rating                159
Review Title          159
Review Text           159
Date of Experience    267
dtype: int64

### 3.3 Duplicate Rows

In [150]:
df.duplicated().sum()

np.int64(2)

### 3.4 Rating Distribution

In [151]:
df['Rating'].value_counts().sort_index()

Rating
Rated 1 out of 5 stars    13123
Rated 2 out of 5 stars     1227
Rated 3 out of 5 stars      885
Rated 4 out of 5 stars     1292
Rated 5 out of 5 stars     4528
Name: count, dtype: int64

### 3.5 Review Text Sample 

In [152]:
df[['Review Title', 'Review Text', 'Rating']].head(10)

,Review Title,Review Text,Rating
0,A Store That Doesn't Want to Sell Anything,"I registered on the website, tried to order a ...",Rated 1 out of 5 stars
1,Had multiple orders one turned up and…,Had multiple orders one turned up and driver h...,Rated 1 out of 5 stars
2,I informed these reprobates,I informed these reprobates that I WOULD NOT B...,Rated 1 out of 5 stars
3,Advertise one price then increase it on website,I have bought from Amazon before and no proble...,Rated 1 out of 5 stars
4,If I could give a lower rate I would,If I could give a lower rate I would! I cancel...,Rated 1 out of 5 stars
5,"Terrible, I had to hang up and call back",Terrible you get customer service reps that ar...,Rated 1 out of 5 stars
6,Yet again,Amazon has a way of tainting a great product d...,Rated 1 out of 5 stars
7,I love amazon,I love amazon! I use it for half my shopping. ...,Rated 5 out of 5 stars
8,I applied for a job with Amazon,I applied for a job with Amazon. I completed a...,Rated 1 out of 5 stars
9,I had a great experience with their…,I had a great experience with their customer s...,Rated 5 out of 5 stars


# Step 4 — Data Cleaning

### 4.1 Remove Duplicate Rows

In [153]:
df = df.drop_duplicates()

### 4.2 Remove Missing Review Text

In [154]:
df = df.dropna (subset=["Review Text"])

### 4.3 Reset Index

In [155]:
df = df.reset_index(drop=True)

### 4.4 Check Result

In [156]:
df.shape

(21055, 9)

In [157]:
df.isnull().sum()

Reviewer Name           0
Profile Link            0
Country                 1
Review Count            0
Review Date             0
Rating                  0
Review Title            0
Review Text             0
Date of Experience    108
dtype: int64

#  Step 5 — Clean Review Text

We clean the review text by converting it to lowercase and removing URLs, punctuation, numbers, and extra spaces. 

This prepares the text for NLP-based sentiment analysis.

### 5.1 Create Text Cleaning Function

In [158]:
def clean_text(text):
    text = text.lower()
    text = re.sub(r'http\S+|www\S+', '', text)
    text = re.sub(r'[^a-z\s]', '', text)
    text = re.sub(r'\s+', ' ', text).strip()
    return text

### 5.2 Apply Cleaning

In [159]:
df['Clean_Review'] = df['Review Text'].apply(clean_text)

### 5.3 Compare Original vs Cleaned Text

In [160]:
df[['Review Text', 'Clean_Review']].head(10)

,Review Text,Clean_Review
0,"I registered on the website, tried to order a ...",i registered on the website tried to order a l...
1,Had multiple orders one turned up and driver h...,had multiple orders one turned up and driver h...
2,I informed these reprobates that I WOULD NOT B...,i informed these reprobates that i would not b...
3,I have bought from Amazon before and no proble...,i have bought from amazon before and no proble...
4,If I could give a lower rate I would! I cancel...,if i could give a lower rate i would i cancell...
5,Terrible you get customer service reps that ar...,terrible you get customer service reps that ar...
6,Amazon has a way of tainting a great product d...,amazon has a way of tainting a great product d...
7,I love amazon! I use it for half my shopping. ...,i love amazon i use it for half my shopping th...
8,I applied for a job with Amazon. I completed a...,i applied for a job with amazon i completed al...
9,I had a great experience with their customer s...,i had a great experience with their customer s...


 #  Step 6 — VADER Sentiment Analysis

### 6.1 Import VADER

In [161]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer

sia = SentimentIntensityAnalyzer()

### 6.2 Calculate Sentiment Scores

In [162]:
df['Sentiment_Score'] = df['Clean_Review'].apply(
    lambda text: sia.polarity_scores(text)['compound']
)

### 6.3 Create Sentiment Category

In [163]:
def classify_sentiment(score):
    if score >= 0.05:
        return 'Positive'
    elif score <= -0.05:
        return 'Negative'
    else:
        return 'Neutral'

df['Sentiment'] = df['Sentiment_Score'].apply(classify_sentiment)

### 6.4 Check Results


In [164]:
df[['Review Text', 'Sentiment_Score', 'Sentiment']].head(10)

,Review Text,Sentiment_Score,Sentiment
0,"I registered on the website, tried to order a ...",0.9197,Positive
1,Had multiple orders one turned up and driver h...,-0.7096,Negative
2,I informed these reprobates that I WOULD NOT B...,-0.5023,Negative
3,I have bought from Amazon before and no proble...,-0.8947,Negative
4,If I could give a lower rate I would! I cancel...,-0.6126,Negative
5,Terrible you get customer service reps that ar...,0.5267,Positive
6,Amazon has a way of tainting a great product d...,0.8904,Positive
7,I love amazon! I use it for half my shopping. ...,0.8779,Positive
8,I applied for a job with Amazon. I completed a...,-0.4310,Negative
9,I had a great experience with their customer s...,0.7003,Positive


In [165]:
df['Sentiment'].value_counts()

Sentiment
Positive    10108
Negative     9394
Neutral      1553
Name: count, dtype: int64

In [166]:
df['Sentiment'].value_counts(normalize=True).mul(100).round(2)

Sentiment
Positive    48.01
Negative    44.62
Neutral      7.38
Name: proportion, dtype: float64

# Step 7 — Sentiment Distribution Visualization

### 7.1 Create Sentiment Counts

In [167]:
sentiment_counts = df['Sentiment'].value_counts().reset_index()
sentiment_counts.columns = ['Sentiment', 'Count']

sentiment_counts

,Sentiment,Count
0,Positive,10108
1,Negative,9394
2,Neutral,1553


### 7.2 Plotly Bar Chart

In [168]:
fig = px.bar(
    sentiment_counts,
    x='Sentiment',
    y='Count',
    text='Count',
    title='Amazon Reviews Sentiment Distribution',
    labels={
        'Sentiment': 'Sentiment',
        'Count': 'Number of Reviews'
    }
)

fig.update_traces(textposition='outside')

fig.update_layout(
    height=500,
    width=1000,
    xaxis_title='Sentiment',
    yaxis_title='Number of Reviews',
    title_x=0.5
)

fig.show()

**Insight:**

* **Positive** reviews represent the largest sentiment group  **10,108**, 
* closely followed by **negative** reviews  **9394**. 
* and **Neutral** reviews account are **1,553** of the dataset.

## Step 8 — Sentiment Percentage Donut Chart

In [169]:
sentiment_percentage = (
    df['Sentiment']
    .value_counts()
    .reset_index()
)

sentiment_percentage.columns = ['Sentiment', 'Count']

fig = px.pie(
    sentiment_percentage,
    names='Sentiment',
    values='Count',
    hole=0.30,
    title='Amazon Reviews Sentiment Percentage'
)

fig.update_traces(
    textinfo='percent+label',
    hovertemplate='<b>%{label}</b><br>Reviews: %{value}<br>Percentage: %{percent}<extra></extra>'
)

fig.update_layout(
    height=500,
    width=1000,
    title_x=0.5
)

fig.show()

**Insight:**

* **Positive** reviews account for **48.01%**, 
* **negative** reviews for **44.62%**,
* and **neutral** reviews for **7.38%** of all reviews.

---

# Step 9 — Sentiment vs Rating Analysis

### 9.1 Clean the Rating Column

In [170]:
df['Rating_Number'] = df['Rating'].str.extract(r'(\d)').astype(int)

In [171]:
df[['Rating', 'Rating_Number']].head()

,Rating,Rating_Number
0,Rated 1 out of 5 stars,1
1,Rated 1 out of 5 stars,1
2,Rated 1 out of 5 stars,1
3,Rated 1 out of 5 stars,1
4,Rated 1 out of 5 stars,1


### 9.2 Sentiment vs Rating

In [172]:
rating_sentiment = (
    df.groupby(['Rating_Number', 'Sentiment'])
      .size()
      .reset_index(name='Count')
)

rating_sentiment

,Rating_Number,Sentiment,Count
0,1,Negative,8071
1,1,Neutral,683
2,1,Positive,4369
3,2,Negative,602
4,2,Neutral,89
5,2,Positive,536
6,3,Negative,311
7,3,Neutral,108
8,3,Positive,466
9,4,Negative,151


### 9.3 Plotly Chart

In [173]:
fig = px.bar(
    rating_sentiment,
    x='Rating_Number',
    y='Count',
    color='Sentiment',
    barmode='group',
    text='Count',
    title='Sentiment Distribution by Customer Rating',
    labels={
        'Rating_Number': 'Customer Rating',
        'Count': 'Number of Reviews',
        'Sentiment': 'Sentiment'
    }
)

fig.update_traces(textposition='outside')

fig.update_layout(
    height=500,
    width= 1000,
    title_x=0.5,
    xaxis=dict(
        tickmode='linear',
        dtick=1
    )
)

fig.show()

**Insight**

* The analysis shows that 1-star reviews contain the highest number of negative sentiments,
* while 5-star reviews contain the highest number of positive sentiments.
* However, some reviews show a different sentiment from their numerical rating,
* indicating that text-based sentiment and star ratings may not always fully agree.

---

# Step 10 — Emotion Detection

In [174]:
from collections import Counter
import re

# Simple emotion lexicon
emotion_lexicon = {
    'joy': ['love', 'happy', 'great', 'excellent', 'amazing', 'wonderful', 'perfect', 'good'],
    'anger': ['angry', 'hate', 'worst', 'terrible', 'awful', 'frustrated', 'annoying'],
    'sadness': ['sad', 'disappointed', 'unhappy', 'bad', 'poor', 'sorry', 'upset'],
    'fear': ['fear', 'scared', 'afraid', 'worried', 'danger', 'risk', 'unsafe'],
    'surprise': ['surprised', 'unexpected', 'shocked', 'wow', 'suddenly', 'astonished']
}

print("Emotion lexicon created successfully.")

Emotion lexicon created successfully.


In [175]:
def detect_emotions(text):
    words = set(re.findall(r'\b[a-z]+\b', text.lower()))
    
    emotion_counts = {}
    
    for emotion, keywords in emotion_lexicon.items():
        count = sum(1 for word in keywords if word in words)
        emotion_counts[emotion] = count
    
    return emotion_counts


df['Emotion_Scores'] = df['Clean_Review'].apply(detect_emotions)

df[['Clean_Review', 'Emotion_Scores']].head()

,Clean_Review,Emotion_Scores
0,i registered on the website tried to order a l...,"{'joy': 0, 'anger': 0, 'sadness': 0, 'fear': 0..."
1,had multiple orders one turned up and driver h...,"{'joy': 0, 'anger': 1, 'sadness': 0, 'fear': 0..."
2,i informed these reprobates that i would not b...,"{'joy': 0, 'anger': 0, 'sadness': 0, 'fear': 0..."
3,i have bought from amazon before and no proble...,"{'joy': 1, 'anger': 0, 'sadness': 1, 'fear': 0..."
4,if i could give a lower rate i would i cancell...,"{'joy': 0, 'anger': 0, 'sadness': 0, 'fear': 0..."


In [176]:
emotion_totals = Counter()

for scores in df['Emotion_Scores']:
    for emotion, count in scores.items():
        emotion_totals[emotion] += count

emotion_df = pd.DataFrame(
    emotion_totals.items(),
    columns=['Emotion', 'Count']
).sort_values('Count', ascending=False)

emotion_df

,Emotion,Count
0,joy,6553
2,sadness,3291
1,anger,2520
4,surprise,352
3,fear,188


### Step 10.4 — Emotion Distribution Chart

In [177]:
fig = px.bar(
    emotion_df,
    x='Emotion',
    y='Count',
    text='Count',
    title='Emotion Distribution in Amazon Reviews',
    labels={
        'Emotion': 'Emotion',
        'Count': 'Number of Detected Emotion Words'
    }
)

fig.update_traces(textposition='outside')

fig.update_layout(
    height=600,
    width= 1000,
    title_x=0.5,
    xaxis_title='Emotion',
    yaxis_title='Number of Detected Emotion Words'
)

fig.show()

**Insight**

**Joy** is the most frequently detected emotion with **6,553 occurrences**, 
* followed by **sadness (3,291)** and **anger (2,520)**.
* **Surprise** and **fear** appear much less frequently, with **352 and 188 occurrences**,
* respectively. Overall, the results indicate that customer reviews contain a mix of positive and negative emotional expressions, with joy being the most prominent emotion.

---

## Step 11 — Sentiment by Country

In [178]:
country_sentiment = (
    df.groupby(['Country', 'Sentiment'])
      .size()
      .reset_index(name='Count')
)

country_sentiment.head(10)

,Country,Sentiment,Count
0,AD,Neutral,1
1,AD,Positive,1
2,AE,Negative,48
3,AE,Neutral,12
4,AE,Positive,29
5,AF,Negative,1
6,AG,Positive,1
7,AM,Negative,5
8,AM,Neutral,1
9,AM,Positive,4


### Step 11.1 — Top 10 Countries

In [179]:
top_countries = (
    df['Country']
    .value_counts()
    .head(10)
    .index
)

country_sentiment_top10 = country_sentiment[
    country_sentiment['Country'].isin(top_countries)
]

country_sentiment_top10

,Country,Sentiment,Count
17,AU,Negative,86
18,AU,Neutral,17
19,AU,Positive,83
48,CA,Negative,345
49,CA,Neutral,57
50,CA,Positive,306
73,DE,Negative,58
74,DE,Neutral,26
75,DE,Positive,97
76,DK,Negative,103


In [180]:
fig = px.bar(
    country_sentiment_top10,
    x='Country',
    y='Count',
    color='Sentiment',
    barmode='group',
    text='Count',
    title='Sentiment Distribution by Top 10 Countries',
    labels={
        'Country': 'Country',
        'Count': 'Number of Reviews',
        'Sentiment': 'Sentiment'
    }
)

fig.update_traces(textposition='outside')

fig.update_layout(
    height=500,
    width= 1000,
    title_x=0.5,
    xaxis_title='Country',
    yaxis_title='Number of Reviews'
)

fig.show()

### **Insight:**

* **US:** Positive **4,616** vs Negative **4,118** — feedback is mixed.
* **GB:** Negative **3,435** vs Positive **3,351** — feedback is nearly balanced.
* **India:** Positive **294** vs Negative **263** — slightly more positive feedback.
* **Overall:** Sentiment varies across countries, with **US and GB** having the highest review volumes.

---


# Step 12 — Sentiment Trend Over Time

In [181]:
df['Review Date'] = pd.to_datetime(df['Review Date'], errors='coerce')

df[['Review Date', 'Sentiment']].head()

,Review Date,Sentiment
0,2024-09-16 13:44:26+00:00,Positive
1,2024-09-16 18:26:46+00:00,Negative
2,2024-09-16 21:47:39+00:00,Negative
3,2024-09-17 07:15:49+00:00,Negative
4,2024-09-16 18:37:17+00:00,Negative


### Step 12.1 — Monthly Sentiment Trend

In [182]:
monthly_sentiment = (
    df.groupby([df['Review Date'].dt.to_period('M'), 'Sentiment'])
      .size()
      .reset_index(name='Count')
)

monthly_sentiment['Review Date'] = monthly_sentiment['Review Date'].astype(str)

monthly_sentiment.head(10)

C:\Users\AL HAFIZ COMPUTER\AppData\Local\Temp\ipykernel_35616\1549031975.py:2: UserWarning:

Converting to PeriodArray/Index representation will drop timezone information.



,Review Date,Sentiment,Count
0,2007-08,Positive,1
1,2008-04,Positive,1
2,2008-09,Positive,1
3,2008-12,Positive,1
4,2009-03,Positive,1
5,2009-04,Negative,1
6,2009-04,Positive,1
7,2009-05,Positive,2
8,2009-07,Positive,1
9,2009-08,Positive,2


### Step 12.2 — Review Date Range

In [183]:
print("Earliest Review:", df['Review Date'].min())
print("Latest Review:", df['Review Date'].max())

Earliest Review: 2007-08-27 17:25:01+00:00
Latest Review: 2024-09-17 13:19:27+00:00


### Step 12.3 — 2024 Monthly Sentiment Trend

In [184]:
df_2024 = df[df['Review Date'].dt.year == 2024].copy()

monthly_sentiment_2024 = (
    df_2024.groupby([
        df_2024['Review Date'].dt.to_period('M'),
        'Sentiment'
    ])
    .size()
    .reset_index(name='Count')
)

monthly_sentiment_2024['Review Date'] = (
    monthly_sentiment_2024['Review Date'].astype(str)
)

monthly_sentiment_2024

C:\Users\AL HAFIZ COMPUTER\AppData\Local\Temp\ipykernel_35616\2916362290.py:5: UserWarning:

Converting to PeriodArray/Index representation will drop timezone information.



,Review Date,Sentiment,Count
0,2024-01,Negative,143
1,2024-01,Neutral,21
2,2024-01,Positive,122
3,2024-02,Negative,165
4,2024-02,Neutral,22
5,2024-02,Positive,101
6,2024-03,Negative,149
7,2024-03,Neutral,16
8,2024-03,Positive,126
9,2024-04,Negative,148


### Step 12.4 — Monthly Sentiment plotly Chart

In [185]:
fig = px.line(
    monthly_sentiment_2024,
    x='Review Date',
    y='Count',
    color='Sentiment',
    markers=True,
    title='Monthly Sentiment Trend in Amazon Reviews - 2024',
    labels={
        'Review Date': 'Month',
        'Count': 'Number of Reviews',
        'Sentiment': 'Sentiment'
    }
)

fig.update_layout(
    height=500,
    width= 1000,
    title_x=0.5,
    xaxis_title='Month',
    yaxis_title='Number of Reviews'
)

fig.show()

### **Insight:**

* **Negative:** Increased from **143** reviews in January to a peak of **219** in July.
* **Positive:** Reached its highest level in **July with 161** reviews.
* **Neutral:** Remained relatively low throughout the period, ranging from **16 to 29** reviews.
* **Overall:** Negative sentiment remained higher than positive sentiment in most months of 2024.


------

## **Marketing, Product & Customer Insights**

* **Marketing:** Use positive customer feedback to highlight product strengths in marketing campaigns.
* **Product Development:** Analyze negative feedback to identify product and service improvement areas.
* **Customer Experience:** Focus on issues reflected through **Sadness and Anger** to improve customer satisfaction.
* **Customer Engagement:** Use positive feedback and **Joy** indicators to understand what customers appreciate.



## **Final Recommendations**

* Monitor negative reviews regularly to identify recurring customer concerns.
* Use emotion patterns to understand customer reactions beyond simple positive or negative sentiment.
* Use positive feedback to identify features and services that customers value.
* Track sentiment trends over time to monitor changes in customer satisfaction.

## **Conclusion**

The sentiment analysis shows that **Positive reviews (48.01%)** slightly exceed **Negative reviews (44.62%)**, while Neutral reviews represent **7.38%**. **Joy** was the most frequently detected emotion, followed by **Sadness** and **Anger**. Overall, sentiment analysis and emotion detection provide useful insights into customer opinions and can support data-driven marketing, product improvement, and customer experience decisions.


-----

**thank you**


